File: mergeSort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(int arr[], int l, int m, int r)
{
    int x = m - l + 1;
    int y = r - m;
    int a[x], b[y];
    for (int i = 0; i < x; i++)
    {
        a[i] = arr[i + l];
    }
    for (int i = 0; i < y; i++)
    {
        b[i] = arr[m + 1 + i];
    }
    int i = 0, j = 0, k = l;
    while (i < x && j < y)
    {
        if (a[i] <= b[j])
        {
            arr[k] = a[i];
            i++;
            k++;
        }
        else if (b[j] < a[i])
        {
            arr[k] = b[j];
            k++;
            j++;
        }
    }
    while (i < x)
    {
        arr[k] = a[i];
        k++;
        i++;
    }
    while (j < y)
    {
        arr[k] = b[j];
        k++;
        j++;
    }
}
void mergeSort(int arr[], int l, int r)
{
    if (l < r)
    {
        int m = l + (r - l) / 2;
        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);
        merge(arr, l, m, r);
    }
}

int main()
{
    int a[] = {8, 6, 3, 2, 1};
    mergeSort(a, 0, 4);
    for (int i = 0; i < 5; i++)
    {
        cout << a[i] << " ";
    }
    return 0;
}

File: merge_sort.py
def merge_sort(arr):
    n = len(arr)
    if n > 1:
        r = n//2
        arr1, arr2 = arr[:r], arr[r:]
        merge_sort(arr1)
        merge_sort(arr2)

        i = j = k = 0
        while i < len(arr1) and j < len(arr2):
            if arr1[i] <= arr2[j]:
                arr[k] = arr1[i]
                i += 1
            else:
                arr[k] = arr2[j]
                j += 1
            k += 1
        while i < len(arr1):
            arr[k] = arr1[i]
            i += 1
            k += 1
        while j < len(arr2):
            arr[k] = arr2[j]
            j += 1
            k += 1


data = [-2, 45, 0, 11, -9]

merge_sort(data)
print(data)


# Time Complexity: O(nlogn) -> O(logn) for recursively splitting the arr in 2 halves and for each subarr, O(n) time for merge op.
# Space Complexity: O(n)
